src/app/pages/admin-page/admin-page.component.ts

```diff
@@ -1,10 +1,321 @@
+      this.markers.push(marker);
+    });
+  }
+
+  private onMapClick(e: L.LeafletMouseEvent): void {
+    const { lat } = e.latlng;
+    const { lng } = e.latlng;
+    if (this.marker) {
+      this.map.removeLayer(this.marker);
+    }
+    this.marker = L.marker([lat, lng], {
+      icon: redIcon,
+      draggable: true,
+    }).addTo(this.map);
+    this.stationForm.patchValue({
+      latitude: lat,
+      longitude: lng,
+    });
+  }
+
+  private createRelation(index: number): FormControl {
+    const control = this.formBuilder.control('');
+    const subscription = control.valueChanges.subscribe(() => {
+      this.checkAndAddOrRemoveRelation(index);
+    });
+    this.formGroupSubscriptions.push(subscription);
+
+    return control;
+  }
+
+  private addRelation(index: number) {
+    const control = this.createRelation(index);
+    this.relations.push(control);
+  }
+
+  private deleteRelation(index: number): void {
+    if (this.relations.length > 1) {
+      this.relations.removeAt(index);
+      const subscription = this.formGroupSubscriptions[index];
+      if (subscription) {
+        subscription.unsubscribe();
+      }
+      this.formGroupSubscriptions.splice(index, 1);
+    }
+  }
+
+  private checkAndAddOrRemoveRelation(changedIndex: number) {
+    const allFilled = this.relations.controls.every((control) => {
+      return control.value;
+    });
+    if (allFilled) {
+      this.addRelation(this.relations.controls.length - 1);
+    } else {
+      const emptyControls = this.relations.controls.filter((control) => {
+        return !control.value;
+      });
+      if (emptyControls.length > 1) {
+        this.deleteRelation(changedIndex);
+      }
+    }
+  }
+
+  protected onSubmit(): void {
+    if (this.stationForm.valid) {
+      const formValue = this.stationForm.value as {
+        city: string;
+        latitude: number;
+        longitude: number;
+        relations: number[];
+      };
+      this.store.dispatch(
+        createStation({
+          city: formValue.city,
+          latitude: formValue.latitude,
+          longitude: formValue.longitude,
+          relations: formValue.relations,
+        })
+      );
+      if (this.marker) {
+        this.map.removeLayer(this.marker);
+      }
+      this.resetForm();
+    }
+  }
+
+  private resetForm() {
+    this.stationForm.reset({
+      city: '',
+      latitude: 0,
+      longitude: 0,
+    });
+    this.resetRelations();
+  }
+
+  protected updateOptions() {
+    const selectedValues = this.relations.controls.map((control) => {
+      return control.value;
+    });
+    this.relations.controls.forEach((control) => {
+      control.setValidators([
+        Validators.required,
+        this.forbidSelectedValues(selectedValues),
+      ]);
+      control.updateValueAndValidity();
+    });
+  }
+
+  private resetRelations() {
+    const relationsArray = this.stationForm.get('relations') as FormArray;
+    while (relationsArray.length !== 0) {
```

Comment:
Here can use
while (relationsArray.length)

Reply:
Thanks, I fixed that and found, that I missed unsubscription, when I reset form